lib/events/filesessions/fileasync.go

```diff
@@ -167,8 +172,22 @@ func (u *Uploader) checkSessionError(sessionID session.ID) (bool, error) {
 
 // Serve runs the uploader until stopped
 func (u *Uploader) Serve(ctx context.Context) error {
+	// Check if close operation is already in progress.
+	// We need to do this because Serve is spawned in a goroutine
+	// and Close can be called before Serve starts which ends up in a data
+	// race because Close is waiting for wg to be 0 and Serve is adding to wg.
+	// To avoid this, we check if Close is already in progress and return
+	// immediately. If Close is not in progress, we add to wg under the mutex
+	// lock to ensure that Close can't reach wg.Wait() before Serve adds to wg.
+	u.mu.Lock()
+	if u.isClosing {
+		u.mu.Unlock()
+		return nil
+	}
```

Comment:
I think that select with closeC will work too so we can get rid of mutex and isClosing: 
```suggestion
	select {
	case <-u.closeC:
		return nil
	default:
	}
```

Reply:
I don't think this would work because we need to ensure that close can't reach  `wg.Wait()` before we either exit `Serve` or call `wg.Add(1)`  there.

If the number of CPUs is >2, it's possible for `Close()` to reach `wg.Wait()` and `Serve` to reach `wg.Add` with that approach